### ISSUE TITLE
Router: Support Setting State Object When Navigating

### ISSUE THREAD

**Issue by vsavkin** [angular/angular#10248]
In addition to setting query params, we should be able to set the state object. See here:

https://developer.mozilla.org/en-US/docs/Web/API/History_API

**Comment by stevenrobijns-macadam** [comment]
So what's the status on this? Will this be included in any of the upcoming releases?

**Comment by DzmitryShylovich** [comment]
@stevenrobijns-macadam what is your use case?

**Comment by stevenrobijns-macadam** [comment]
@DzmitryShylovich We want to supply data (for example an id or maybe a whole object) from one screen to another while navigating. BUT without need to put that info in the URI.

So when going from api/order to api/orderline/orderLineId with router.navigate I for example want to supply the OrderId in a hidden manner, i don't want to see that OrderId reflected in the route.

**Comment by stevenrobijns-macadam** [comment]
@DzmitryShylovich Did you see my reply?

**Comment by f-aubert** [comment]
Is there anything new? Passing an object to navigate methods, instead of having to stringify it as queryParams or save it in a Service, would be most valuable. Like others I cannot think of reasons where it would hurt to give front end developer this option.

**Comment by stevenrobijns-macadam** [comment]
I'm still wondering as well.

**Comment by stevenrobijns-macadam** [comment]
https://stackoverflow.com/questions/37157838/angular-2-passing-data-to-routes 
@f-aubert check if Shubham Verma's comment please and keep us up-to-date if it works for you

**Comment by StevenLiekens** [comment]
Guys, passing data via a service will obviously work but it is a cheap trick that comes with pitfalls.

The biggest pitfall is persistence. When you store router state in a shared service, you must write additional code to save that state to local storage (or whatever storage) and even more code to read the state from that storage when the application is restarted. Then you must start thinking about stale data. You probably don't want to keep route data around forever, do you? So you have to write code that cleans up route data when it's no longer needed. Turns out it's buggy so you promise your manager you'll fix it in the next release. So you fix it in between building other features and then you realize you must think about versioning the data because the new code is not compatible with the data that users have in their cache. Now is the time when you realize you're in hell and you wish you became a carpenter instead.

If you don't do all of this, your app will break if the user navigates to your routable component in a way that bypasses the procedure where you set the state, such as when the page is refreshed.

So let's not ever again bring up the suggestion that you can pass data via a service and instead let's start talking about how we can design the router to take advantage of state in the history API.

**Comment by stevenrobijns-macadam** [comment]
@StevenLiekens: we're not really talking about saving state, that's a whole other issue.
The question is how we can get data from component a to component b. Let's say how can we pass an 'Order' object from the 'OrderComponent' to the 'OrderDetailComponent'. In this case, state should only be saved during the routing between 2 components. And not during application restarts.

**Comment by StevenLiekens** [comment]
@stevenrobijns-macadam The thing is that with the history's `pushState()` API, that state is persisted automatically. That means that when you refresh the page or re-visit it from history, your component would be reloaded with the exact same state. You cannot get this behavior without writing a LOT of custom code if you use a shared service for passing state to a route.

**Comment by f-aubert** [comment]
@stevenrobijns-macadam thanks for the link although it's not a solution for me as my use case would require to dynamically pass data. I know there are many ways of landing on a given route / component, but a common use case is the list of objects and then the object details. As angular is after all mostly  a one page application, there should be the possibility to pass data in memory for a specific routing resolution. As other said service could work but then we would have to manually clean such one time registered objects

**Comment by stevenrobijns-macadam** [comment]
@f-aubert you're right, I didn't look into it much. This guy just proposes a solution for passing data(a string) dynamically in the url, which is not our question.

**Comment by sashanana** [comment]
Any update on this?

**Comment by RenaldasK** [comment]
+1

**Comment by saeedhomsy** [comment]
Is there any update

**Comment by StevenLiekens** [comment]
Can has update pl0x

**Comment by StevenLiekens** [comment]
This is still a huge pain to deal with. Shared services only get you so far. They are not a good substitute for `history.state`.

The best thing I could come up with is a service that controls navigation and also acts as resolve guard so that state is attached to every route snapshot.

```ts
class StatefulNavigationService implements Resolve<any> {
  private state: any;
  
  constructor(private router: Router) {}
  
  navigateWithState(commands: any[], extras?: NavigationExtras, state?: any) {
    this.state = state;
    this.router.navigate(commands, extras).then(() => this.state = null);
  }
  
  resolve(route: ActivatedRouteSnapshot, routerState: RouterStateSnapshot): any {
    return this.state;
  }
}
```

**Comment by suraj021** [comment]
It's Aug'18, any update on this. This issue is pending from 2016.

**Comment by andreaslarssen** [comment]
+1

**Comment by fredrik-macrobond** [comment]
+1

**Comment by Virendra-Yadav** [comment]
you guys can use https://www.npmjs.com/package/ngx-navigation-with-data

**Comment by Virendra-Yadav** [comment]
> +1

you guys can use https://www.npmjs.com/package/ngx-navigation-with-data

**Comment by jasonaden** [comment]
Closed by #27198
